### PR TITLE
support e4m3 kvcache in qwen2 & add kv scaling facotr json

### PR DESCRIPTION
**Pull request by bjmsong**
## Modifications

follow up https://github.com/sgl-project/sglang/pull/2786
- support fp8(e4m3) kv cache in qwen2
- add kv scaling facotr json
- add test
- add benchmark result

## Checklist

- [ ] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contribution_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contribution_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

**Comment by ispobock** [comment]
Hi @bjmsong ref https://github.com/sgl-project/sglang/pull/2786#issuecomment-2591381484, could you provide some throughput/latency related benchmark data with bench_serving on different request rates?

**Comment by merrymercy** [comment]
Assigned to @ispobock

**Comment by bjmsong** [comment]
on A100(40GB)
| request-rate | KV Cache | throughput (tok/s) | Median TTFT (ms) | Median TPOT (ms) |
| ------------ | -------- | ------------------ | ---------------- | ---------------- |
| 32           | FP8-E5M2 | 3640               | 77               | 31               |
|              | FP8-E4M3 | 1527               | 112              | 83               |


- The performance is weird. But after set `export SGLANG_FLASHINFER_USE_TENSOR_CORE=True`, which related to this [PR](https://github.com/sgl-project/sglang/pull/2496), the performance looks normal:

| request-rate | KV Cache | throughput (tok/s) | Median TTFT (ms) | Median TPOT (ms) |
| ------------ | -------- | ------------------ | ---------------- | ---------------- |
| 32           | FP8-E5M2 | 4497               | 76               | 24               |
|              | FP8-E4M3 | 4474               | 70               | 24               |
| 64           | FP8-E5M2 | 4710               | 87               | 23               |
|              | FP8-E4M3 | 4699               | 82               | 23               |

File: test/srt/test_fp8_kvcache.py [hunk 1]
```diff
@@ -18,7 +18,7 @@ def setUpClass(cls):
         cls.model = DEFAULT_MODEL_NAME_FOR_TEST
         cls.base_url = DEFAULT_URL_FOR_TEST
         dirpath = os.path.dirname(__file__)
-        config_file = os.path.join(dirpath, "kv_cache_scales_llama3_8b_chat.json")
```
Comment (ispobock) [inline test/srt/test_fp8_kvcache.py:21]: Add test cases for both llama and qwen.
Add this file to [run_suite.py](https://github.com/sgl-project/sglang/blob/main/test/srt/run_suite.py).

File: python/sglang/srt/models/qwen2.py [hunk 2]
```diff
@@ -20,8 +20,12 @@
 
 import torch
 from torch import nn
-from vllm.distributed import get_tensor_model_parallel_world_size
+from vllm.distributed import (
+    get_tensor_model_parallel_rank,
+    get_tensor_model_parallel_world_size,
+)
 from vllm.model_executor.layers.rotary_embedding import get_rope
+from vllm.model_executor.model_loader.weight_utils import kv_cache_scales_loader
```
Comment (ispobock) [inline python/sglang/srt/models/qwen2.py]: seems a new dependency, maybe we can replace it with code
cc:@zhyncs

**Review by ispobock — CHANGES_REQUESTED** [review]

**Comment by ispobock** [comment]
- How about the bench serving performance compared to FP16?
- A100(SM80) doesn't support FP8 tensor core. Why `export SGLANG_FLASHINFER_USE_TENSOR_CORE=True` can get acceleration? Did you verified the evaluation accuracy?

File: python/sglang/srt/models/qwen2.py (hunk 2 above)
Comment (bjmsong) [reply python/sglang/srt/models/qwen2.py]: it has been replaced

**Comment by bjmsong** [comment]
> How about the bench serving performance compared to FP16?
- `SGLANG_FLASHINFER_USE_TENSOR_CORE=True`

| request-rate | KV Cache | throughput (tok/s) | Median TTFT (ms) | Median TPOT (ms) |
| ------------ | -------- | ------------------ | ---------------- | ---------------- |
| 32           | FP16     | 4058               | 1046             | 26               |
|              | FP8-E5M2 | 4169               | 1171             | 25               |
|              | FP8-E4M3 | 4226               | 1057             | 25               |

> A100(SM80) doesn't support FP8 tensor core. Why export SGLANG_FLASHINFER_USE_TENSOR_CORE=True can get acceleration? Did you verified the evaluation accuracy?
- It uses bf16 tensor cores
- Accuracy test

| KV Cache | use tensor core | mmlu  | mgsm_en |
| -------- | --------------- | ----- | ------- |
| FP8-E4M3 | True            | 0.625 | 0.796   |
| FP8-E4M3 | False           | 0.625 | 0.800   |

File: test/srt/test_fp8_kvcache.py (hunk 1 above)
Comment (bjmsong) [reply test/srt/test_fp8_kvcache.py:21]: done

**Review by ispobock — APPROVED** [review]

**Comment by jjjjohnson** [comment]
@bjmsong Thans for the great work! How to generate the scaling factor json file? Looks like llmcompressor do not have this function.

**Comment by bjmsong** [comment]
> @bjmsong Thans for the great work! How to generate the scaling factor json file? Looks like llmcompressor do not have this function.

The model weight compressed by llmcompressor contains the scaling factor, whose name contains `kv_scale`.
